components/raftstore-v2/src/operation/command/admin/merge/commit.rs

```diff
@@ -0,0 +1,652 @@
+                "source" => ?source_region,
+                "target" => ?region,
+            );
+            Some(false)
+        } else if expected_epoch == region.get_region_epoch() {
+            Some(true)
+        } else {
+            None
+        }
+    }
+
+    pub fn propose_commit_merge<T>(
+        &mut self,
+        store_ctx: &mut StoreContext<EK, ER, T>,
+        req: RaftCmdRequest,
+    ) -> Result<u64> {
+        let mut proposal_ctx = ProposalContext::empty();
+        proposal_ctx.insert(ProposalContext::COMMIT_MERGE);
+        let data = req.write_to_bytes().unwrap();
+        self.propose_with_ctx(store_ctx, data, proposal_ctx.to_vec())
+    }
+}
+
+impl<EK: KvEngine, R: ApplyResReporter> Apply<EK, R> {
+    // Match v1::exec_commit_merge.
+    pub async fn apply_commit_merge(
+        &mut self,
+        req: &AdminRequest,
+        index: u64,
+    ) -> Result<(AdminResponse, AdminCmdResult)> {
+        PEER_ADMIN_CMD_COUNTER.commit_merge.all.inc();
+
+        self.flush();
+
+        // Note: compared to v1, doesn't validate region state from kvdb any more.
+        let reg = self.tablet_registry();
+        let merge = req.get_commit_merge();
+        let source_region = merge.get_source();
+        let source_path = merge_source_path(reg, source_region.get_id(), merge.get_commit());
+        if !source_path.exists() {
+            let (tx, rx) = oneshot::channel();
+            self.res_reporter()
+                .redirect_catch_up_logs(self.region_id(), merge.clone(), tx);
+            rx.await.unwrap_or_else(|_| {
+                // TODO: handle this gracefully.
+                slog_panic!(self.logger, "source peer is missing");
+            });
+        }
+        let ctx = TabletContext::new(source_region, None);
+        let source_tablet = reg.tablet_factory().open_tablet(ctx, &source_path).unwrap();
+
+        info!(
+            self.logger,
+            "execute CommitMerge";
+            "commit" => merge.get_commit(),
+            "entries" => merge.get_entries().len(),
+            "index" => index,
+            "source_region" => ?source_region,
+        );
+        let mut region = self.region().clone();
+        // Use a max value so that pd can ensure overlapped region has a priority.
+        let version = cmp::max(
+            source_region.get_region_epoch().get_version(),
+            region.get_region_epoch().get_version(),
+        ) + 1;
+        region.mut_region_epoch().set_version(version);
+        if keys::enc_end_key(&region) == keys::enc_start_key(source_region) {
+            region.set_end_key(source_region.get_end_key().to_vec());
+        } else {
+            region.set_start_key(source_region.get_start_key().to_vec());
+        }
+
+        // TODO: check both are trimmed.
+
+        let path = reg.tablet_path(self.region_id(), index);
+        // The last merge is incomplete.
+        if path.exists() {
+            std::fs::remove_dir_all(&path).unwrap();
```

Comment:
It may be wrong to delete it directly. Because during recovery, modifications to some CF can be skipped if they are flushed in the past.

Reply:
Should explain why it's safe to delete the directory.